sw/device/tests/otbn_mem_scramble_test.c

```diff
@@ -0,0 +1,110 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_base.h"
+#include "sw/device/lib/dif/dif_otbn.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/otbn.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTBN_DECLARE_APP_SYMBOLS(randomness);
+static const otbn_app_t kOtbnAppCfiTest = OTBN_APP_T_INIT(randomness);
+
+const test_config_t kTestConfig;
+
+static volatile bool has_exception_fired;
+
+/**
+ * This overrides the default OTTF load/store fault exception handler.
+ */
+void ottf_load_store_fault_handler(void) { has_exception_fired = true; }
+
+typedef dif_result_t (*otbn_read_t)(const dif_otbn_t *otbn,
+                                    uint32_t offset_bytes, void *dest,
+                                    size_t len_bytes);
+/**
+ * Check whether the contents of an otbn memory match the reference data pointed
+ * at `addr`.
+ *
+ * @param ctx The otbn context object.
+ * @param addr The address of the reference data.
+ * @param mem_size The size of the reference data.
+ * @param match_expected Indicates whether the checking is expected to match.
+ * @param read_function Pointer to the function to read the data. It can be
+ * either `dif_otbn_imem_read` or `dif_otbn_dmem_read`.
+ */
+static void otbn_check_mem(otbn_t *ctx, const uint8_t *addr, size_t mem_size,
+                           bool match_expected, otbn_read_t otbn_read) {
+  uint8_t local_buf[256];
+  size_t offset = 0;
+  do {
+    size_t remainder = mem_size - offset;
+    if (remainder > ARRAYSIZE(local_buf)) {
+      remainder = ARRAYSIZE(local_buf);
+    }
+
+    // If the memory has been scrambled we will expect to receive an exception,
+    // otherwise we compare the memory value.
+    has_exception_fired = false;
+    CHECK_DIF_OK(otbn_read(&ctx->dif, offset, local_buf, remainder));
+    if (match_expected) {
+      CHECK(has_exception_fired == false, "Unexpected exception");
+      CHECK_BUFFER(addr + offset, local_buf, remainder);
+    } else {
+      CHECK(has_exception_fired, "Expected exception haven't fired");
```

Comment:
Nit: This should probably be "hasn't" (because there's only one expected exception).